### PR TITLE
Update scoping with latest upstream changes

### PR DESCRIPTION
**Pull request by cbeer**

### CODE REVIEW

File: lib/active_fedora/identifiable.rb [hunk 1]
```diff
@@ -83,6 +83,28 @@ def from_uri(uri, _)
...
+      # Returns the class descending directly from ActiveRecord::Base, or
+      # an abstract class, if any, in the inheritance hierarchy.
+      #
+      # If A extends ActiveRecord::Base, A.base_class will return A. If B descends from A
+      # through some arbitrarily deep hierarchy, B.base_class will return A.
+      #
+      # If B < A and C < B and if A is an abstract_class then both B.base_class
+      # and C.base_class would return B as the answer since A is an abstract_class.
+      def base_class
+        return File if self == File || superclass == File
```
Comment (cbeer) [inline lib/active_fedora/identifiable.rb]: Something's not quite right about our class hierarchy, and there's also something funny about our associations that tries to get the base class for`AF::Base`.

**Comment by jcoyne** [comment]
Rubocop isn't happy with the style

**Comment by jcoyne** [comment]
Is this ready now? If so, can you remove the "[WIP]" ?

File: lib/active_fedora/scoping/named.rb [hunk 2]
```diff
@@ -34,6 +34,146 @@ def default_scoped
...
+        # <tt>Shirt.red</tt> really was an array.
+        #
+        # These named \scopes are composable. For instance,
+        # <tt>Shirt.red.dry_clean_only</tt> will produce all shirts that are
+        # both red and dry clean only. Nested finds and calculations also work
+        # with these compositions: <tt>Shirt.red.dry_clean_only.count</tt>
+        # returns the number of garments for which these criteria obtain.
+        # Similarly with <tt>Shirt.red.dry_clean_only.average(:thread_count)</tt>.
+        #
+        # All scopes are available as class methods on the ActiveRecord::Base
```
Comment (jcoyne) [inline lib/active_fedora/scoping/named.rb]: Remove ActiveRecord and replace with ActiveFedora

File: lib/active_fedora/inheritance.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+module ActiveFedora
+  module Inheritance
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      # Returns the class descending directly from ActiveRecord::Base, or
```
Comment (jcoyne) [inline lib/active_fedora/inheritance.rb]: Replace ActiveRecord with ActiveFedora

File: lib/active_fedora/inheritance.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      # an abstract class, if any, in the inheritance hierarchy.
+      #
+      # If A extends ActiveRecord::Base, A.base_class will return A. If B descends from A
```
Comment (jcoyne) [inline lib/active_fedora/inheritance.rb]: Replace ActiveRecord with ActiveFedora

File: lib/active_fedora/scoping.rb [hunk 5]
```diff
@@ -20,5 +29,74 @@ def scope_attributes? # :nodoc:
...
+    # This class stores the +:current_scope+ and +:ignore_default_scope+ values
+    # for different classes. The registry is stored as a thread local, which is
+    # accessed through +ScopeRegistry.current+.
+    #
+    # This class allows you to store and get the scope values on different
+    # classes and different types of scopes. For example, if you are attempting
+    # to get the current_scope for the +Board+ model, then you would use the
+    # following code:
+    #
+    #   registry = ActiveRecord::Scoping::ScopeRegistry
```
Comment (jcoyne) [inline lib/active_fedora/scoping.rb]: Again

File: lib/active_fedora/scoping.rb [hunk 6]
```diff
@@ -20,5 +29,74 @@ def scope_attributes? # :nodoc:
...
+    #
+    # Now when you run:
+    #
+    #   registry.value_for(:current_scope, Board)
+    #
+    # You will obtain whatever was defined in +some_new_scope+. The #value_for
+    # and #set_value_for methods are delegated to the current ScopeRegistry
+    # object, so the above example code can also be called as:
+    #
+    #   ActiveRecord::Scoping::ScopeRegistry.set_value_for(:current_scope,
```
Comment (jcoyne) [inline lib/active_fedora/scoping.rb]: And here

File: lib/active_fedora/scoping/named.rb [hunk 7]
```diff
@@ -34,6 +34,146 @@ def default_scoped
             relation
           end
         end
+
+        # Adds a class method for retrieving and querying objects.
+        # The method is intended to return an ActiveRecord::Relation
```
Comment (jcoyne) [inline lib/active_fedora/scoping/named.rb]: Here too.

File: lib/active_fedora/scoping/named.rb [hunk 8]
```diff
@@ -34,6 +34,146 @@ def default_scoped
...
+        # class method:
+        #
+        #   class Shirt < ActiveFedora::Base
+        #     def self.red
+        #       where(color: 'red')
+        #     end
+        #   end
+        #
+        # Unlike <tt>Shirt.find(...)</tt>, however, the object returned by
+        # <tt>Shirt.red</tt> is not an Array but an ActiveRecord::Relation,
```
Comment (jcoyne) [inline lib/active_fedora/scoping/named.rb]: Ditto

**Comment by cbeer** [comment]
Still WIP. I still need to track down why I had to add some hacks to `.base_class`

**Comment by jcoyne** [comment]
@cbeer I'm happy to pair on it with you if you want a hand.

**Comment by cbeer** [comment]
:+1:

**Pull request merged**